joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -165,32 +149,19 @@ JointTrajectoryController::update()
 
   // current state update
   state_current.time_from_start.set__sec(0);
-
-  // Assign values from the hardware
-  // Position states always exist
-  assign_point_from_interface(state_current.positions, joint_state_interface_[0]);
-  // velocity and acceleration states are optional
-  if (has_velocity_state_interface_) {
-    assign_point_from_interface(state_current.velocities, joint_state_interface_[1]);
-    // Acceleration is used only in combination with velocity
-    if (has_acceleration_state_interface_) {
-      assign_point_from_interface(state_current.accelerations, joint_state_interface_[2]);
-    } else {
-      // Make empty so the property is ignored during interpolation
-      state_current.accelerations.clear();
-    }
-  } else {
-    // Make empty so the property is ignored during interpolation
-    state_current.velocities.clear();
-    state_current.accelerations.clear();
-  }
+  read_state_from_hardware(state_current);
 
   // currently carrying out a trajectory
   if (traj_point_active_ptr_ && (*traj_point_active_ptr_)->has_trajectory_msg()) {
     // if sampling the first time, set the point before you sample
     if (!(*traj_point_active_ptr_)->is_sampled_already()) {
-      (*traj_point_active_ptr_)->set_point_before_trajectory_msg(
-        node_->now(), state_current);
+      if (open_loop_control_) {
+        (*traj_point_active_ptr_)->set_point_before_trajectory_msg(
+          node_->now(), last_commanded_state_);
```

Comment:
Should `last_commanded_state_` here be reduced to existing state interface from the hardware? Or is it OK to internally manage more control-states then can be received from hardware.

Reply:
I think this should only set the ones you'd see from commands too, otherwise may result in strange discrepancies. Happy to have a follow-up issue for it